### PR TITLE
archive node retries

### PR DESCRIPTION
**Pull request by deepthiskumar**
Added retry logic for database writes to support concurrent database transactions (described [here](https://github.com/MinaProtocol/mina/issues/7567)). This also requires setting transaction isolation level to 'serializable' and can be done in two ways:

1. set it as the default isolation level in postgresql.conf:
     `default_transaction_isolation = 'serializable'`
2. set it for a specific database using the following query
    `ALTER DATABASE <DATABASE NAME> SET DEFAULT_TRANSACTION_ISOLATION TO SERIALIZABLE ;` 

They need to be done as a separate step before archive node starts writing (and before bringing up the database instance in the case of option 1). I think setting it at the db level makes it explicit and therefore is a better option, but i'm not quite sure where in the codebase this configuration needs to live

### CODE REVIEW

**Comment by psteckler** [comment]
Can you confirm that `make archive_blocks` still builds (there are additional arguments to the `make_block_*_aux` functions)?

File: src/app/archive/archive_lib/processor.ml [hunk 1]
```diff
@@ -1355,34 +1355,50 @@ module Block = struct
...
-    | Some num_blocks ->
-        Block.delete_if_older_than ~num_blocks (module Conn)
-    | None ->
-        return ()
-  in
-  let%map () =
     match res with
-    | Error _ ->
-        Conn.rollback () >>| ignore
+    | Error _ as e ->
```
Comment (psteckler) [inline src/app/archive/archive_lib/processor.ml:1394]: Should there be logs for the commit and rollback?

Comment (deepthiskumar) [reply src/app/archive/archive_lib/processor.ml]: yup, added it

**Comment by deepthiskumar** [comment]
> Can you confirm that `make archive_blocks` still builds (there are additional arguments to the `make_block_*_aux` functions)?



> Can you confirm that `make archive_blocks` still builds (there are additional arguments to the `make_block_*_aux` functions)?

There was a type error, fixed it

File: src/app/archive/archive_lib/processor.ml [hunk 2]
```diff
@@ -1355,34 +1355,57 @@ module Block = struct
     else return ()
 end
 
-let add_block_aux ~add_block ~hash ~delete_older_than
+let retry ~f ~logger ~error_str retries =
+  let rec go retry_count =
+    match%bind f () with
+    | Error e ->
+        if retry_count <= 0 then return (Error e)
```
Comment (psteckler) [inline src/app/archive/archive_lib/processor.ml:1362]: maybe accumulate errors?

Comment (deepthiskumar) [reply src/app/archive/archive_lib/processor.ml:1362]: In addition to logging it as warnings (the else case)? The error in this line is logged at the call site

**Review by psteckler — APPROVED** [review]

File: src/app/archive/archive_lib/processor.ml (hunk 2 above)
Comment (psteckler) [reply src/app/archive/archive_lib/processor.ml:1362]: probably good enough

File: src/app/archive/archive_lib/processor.ml [hunk 3]
```diff
@@ -1355,34 +1355,57 @@ module Block = struct
...
+        else (
+          [%log warn] "Error in %s : $error. Retrying..." error_str
+            ~metadata:[("error", `String (Caqti_error.show e))] ;
+          go (retry_count - 1) )
```
Comment (bkase) [inline src/app/archive/archive_lib/processor.ml:1366]: do you want to immediately retry without any backoff timeout? Maybe it's worth waiting a few milliseconds? Though I'm not sure what else that will affect. Up to you...

**Review by bkase — APPROVED** [review]

File: src/app/archive/archive_lib/processor.ml (hunk 3 above)
Comment (deepthiskumar) [reply src/app/archive/archive_lib/processor.ml:1366]: Don't see a problem adding a backoff timeout of a few milliseconds. It should help if there are multiple commit attempts

**Comment by 0x0I** [comment]
>     2. set it for a specific database using the following query
>        `ALTER DATABASE <DATABASE NAME> SET DEFAULT_TRANSACTION_ISOLATION TO SERIALIZABLE ;`
> They need to be done as a separate step before archive node starts writing (and before bringing up the database instance in the case of option 1). I think setting it at the db level makes it explicit and therefore is a better option, but i'm not quite sure where in the codebase this configuration needs to live

Agreed and something like below added to archive node's db-bootstrap [jobs](https://github.com/MinaProtocol/mina/blob/develop/helm/archive-node/templates/db-bootstrap.yaml#L8) should work for this:
```
      - name: update-transaction-isolation-level
        image: bitnami/postgresql
        command: ["bash", "-c"]
        args: ["PGPASSWORD={{ .Values.postgresql.postgresqlPassword }} psql --username {{ .Values.postgresql.postgresqlUsername }} --host {{ tpl .Values.archive.postgresHost . }} --port {{ .Values.archive.postgresPort }} --dbname {{ .Values.archive.postgresDB}} -c ALTER DATABASE {{ .Values.archive.postgresDB }} SET DEFAULT_TRANSACTION_ISOLATION TO SERIALIZABLE ;"]
```

**Comment by 0x0I** [comment]
>     1. set it as the default isolation level in postgresql.conf:
>        `default_transaction_isolation = 'serializable'

Alternatively, we can update the postgres [settings](https://github.com/MinaProtocol/mina/blob/develop/automation/terraform/modules/kubernetes/testnet/helm.tf#L123) set in terraform with extended configs like:

```
postgresqlExtendedConf = {
      default_transaction_isolation ="serializable"
}
```

**Comment by aneesharaines** [comment]
!approved-for-mainnet

**Pull request merged**